WEB-INTEGRATION.md

```diff
@@ -0,0 +1,742 @@
+      will synchronously fire a `click` event
+    - Setting [`location.hash`](https://html.spec.whatwg.org/multipage/nav-history-apis.html#dom-location-hash)
+      will synchronously cause a same-document navigation which fires a
+      `popstate` event on the window object.
+
+  * In cases where a web API queues a task to fire an event, or starts an
+    asynchronous or parallel operation that will ultimately queue such a task,
+    the context in which that web API is called is the **async dispatch
+    context** for that event. (If there is a sync dispatch context, that usually
+    takes priority over an async context.)
+
+    Examples of cases where there would be an async dispatch context:
+    - [`xhr.send()`](https://xhr.spec.whatwg.org/#dom-xmlhttprequest-send),
+      which starts an asynchronous fetch which will eventually fire the various
+      XHR events (e.g. `load`).
+    - Setting `img.src`, which starts an asynchronous image load which will
+      eventually fire the `load` event.
+
+  In cases where an event is caused by a user interaction, or otherwise
+  triggered by the browser, there is no dispatch context, whether sync or async.
+  Examples are:
+  - A `click` event caused by a user click.
+  - The `offline` and `online` events on the window object, which are fired when
+    the browser detects a change of network connectivity.
+  - Events caused by JS code outside the agent, such as a `message` event coming
+    from a worker.
+
+When an event listener is added, the registration context will be saved along
+with the listener, and when the event is dispatched, each listener will be
+called in its registration context. This fits the general scheme of storing the
+snapshot when the callback is passed to a web API, and running the callback in
+that snapshot.
+
+However, for certain events, the dispatch context might be important for
+authors. In those cases, we propose extending specific Event subclasses with a
+property containing an `AsyncContext.Snapshot` object. In cases where there
+might or might not be a dispatch snapshot (e.g. the `click` event, which would
+have a dispatch snapshot if it originates from `el.click()` but not if it
+originates from a user click), this property would be `null` if there isn’t one.
+Note that the web platform can start out conservatively adding these properties
+only for certain `Event` subclasses, and incrementally add support for more
+later.
+
+For `Event` subclasses in which this property is nullable, constructing them
+from JavaScript should set that property to `null`. It might be possible to
+extend the constructor’s options bag to allow capturing the snapshot at
+construction time, but this would not be part of the initial rollout. For
+subclasses where the property is not nullable, calling the constructor could
+capture the snapshot. Another possibility is to always have this property as
+nullable, even for event types where there is always a dispatch context.
```

Comment:
I would lean toward instead letting the developer set it to any snapshot value. The general pattern for event constructors is to let people set whatever they want for synthetic events.